### PR TITLE
Add support for camera and new land pad in irlock simulation

### PR DESCRIPTION
**Pull request by JonasPerolini**
This pull request extends the current irlock simulation `make px4_sitl gazebo_iris_irlock` to support testing of the changes proposed in [another pull request](https://github.com/PX4/PX4-Autopilot/pull/20653) that aims to improve the `landing_target_estimator` module by: 

1. fusing several measurements (mainly irlock, GPS on target, and vision-based fiducial marker detection) in a Kalman Filter
2. estimating the full pose of static and moving targets. 

The following modifications have been made:

- Equipped the Iris drone with a camera with a new aruco marker detection plugin using _open-cv_.
- Added a new land pad with:
    - An Aruco marker
    - A GPS (sends position and velocity of the target)
    - An irlock (simulated as a [logical camera](https://classic.gazebosim.org/tutorials?tut=logical_camera_sensor&cat=sensors))
    - Velocity plugin to simulate moving targets (The speed can be changed in _models/land_pad/model.sdf_)
    
The new land pad and plugins will allow for testing of the changes proposed in the other pull request in a realistic simulation environment. The code has been tested (see examples below) and is ready for review. Please let me know if there are any further modifications needed.

Thanks!

Tests:
- Landing using only irlock: ok
- Landing using only target GPS: ok
- Landing using only vision-based marker detection: ok
- Target moving: ok

Results show that each sensor observation is correct as the drone is able to land precisely on the target.

Example fusing Irlock, GPS and vision: 

https://user-images.githubusercontent.com/74473718/212858668-53fe5220-db7f-4aa9-b3f3-671221d090bf.mp4

prec_land_static_target

Note: the Kalman filter estimates the target GPS bias, this explains why there is a slight difference between the target GPS observation and the target estimated state. 

Example moving target: 

https://user-images.githubusercontent.com/74473718/212861990-5209bb92-004f-4bb2-b15f-d7ad6d5ea103.mp4

Note: The drone does not land on the moving target because the current PX4 precision landing algorithm does not support moving targets. The drone is thus stuck in horizontal approach. 

Moving_target

### CODE REVIEW

**Comment by hamishwillee** [comment]
@Jaeyoung-Lim @TSC21 Who's the best person to review these? What level of support is being offered for gazebo classic?

**Comment by Jaeyoung-Lim** [comment]
@hamishwillee This is on me,

Gazebo classic is the default and main simulator :)

File: src/gazebo_mavlink_interface.cpp [hunk 1]
```diff
@@ -957,6 +962,65 @@ void GazeboMavlinkInterface::IRLockCallback(IRLockPtr& irlock_message) {
...
+  sensor_msg.yaw_var = arucoMarker_message->yaw_var();
+
+  mavlink_message_t msg;
+  mavlink_msg_target_relative_encode_chan(1, 200, MAVLINK_COMM_0, &msg, &sensor_msg);
+  mavlink_interface_->send_mavlink_message(&msg);
+}
+
+void GazeboMavlinkInterface::TargetGpsCallback(GpsPtr& gps_msg) {
+
+  mavlink_target_absolute_t gps_data;
```
Comment (Jaeyoung-Lim) [inline src/gazebo_mavlink_interface.cpp:999]: These mavlink messages are not merged (https://github.com/mavlink/mavlink/pull/1943) Therefore, I don't think we can merge this PR into this repository

File: src/gazebo_mavlink_interface.cpp [hunk 2]
```diff
@@ -957,6 +962,65 @@ void GazeboMavlinkInterface::IRLockCallback(IRLockPtr& irlock_message) {
   mavlink_interface_->send_mavlink_message(&msg);
 }
 
+void GazeboMavlinkInterface::ArucoMarkerCallback(ArucoMarkerPtr& arucoMarker_message) {
+  
+  mavlink_target_relative_t sensor_msg;
```
Comment (Jaeyoung-Lim) [inline src/gazebo_mavlink_interface.cpp:967]: These mavlink messages are not merged (https://github.com/mavlink/mavlink/pull/1943) Therefore, I don't think we can merge this PR into this repository

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]

File: msgs/SITLGps.proto [hunk 3]
```diff
@@ -14,4 +14,9 @@ message SITLGps
   optional double velocity_east         = 9;
   optional double velocity_north        = 10;
   optional double velocity_up           = 11;
+  optional double gt_noise_gps_eph      = 12;
+  optional double gt_noise_gps_epv      = 13;
+  optional double gt_noise_gps_vel_x    = 14;
+  optional double gt_noise_gps_vel_y    = 15;
+  optional double gt_noise_gps_vel_z    = 16;
```
Comment (Jaeyoung-Lim) [inline msgs/SITLGps.proto]: How are these noise information being used?

File: src/gazebo_mavlink_interface.cpp (hunk 1 above)
Comment (potaito) [reply src/gazebo_mavlink_interface.cpp:999]: Give it a few more days :smile:

Comment (Jaeyoung-Lim) [reply src/gazebo_mavlink_interface.cpp:999]: @potaito Sure :grin:

File: src/gazebo_irlock_plugin.cpp [hunk 4]
```diff
@@ -80,27 +83,64 @@ void IRLockPlugin::Load(sensors::SensorPtr _sensor, sdf::ElementPtr _sdf)
...
 
+        /* Save the attitude of the drone when the frame was grabbed, will allow to transform from FRD to vc-NED */
+        ignition::math::Quaterniond q_FLU_to_NED = q_ENU_to_NED * att_W_I;
+        ignition::math::Quaterniond q_nb = q_FLU_to_NED * q_FLU_to_FRD.Inverse();
+
+        irlock_message.set_attitude_q_w(q_nb.W());
+        irlock_message.set_attitude_q_x(q_nb.X());
+        irlock_message.set_attitude_q_y(q_nb.Y());
+        irlock_message.set_attitude_q_z(q_nb.Z());
+
```
Comment (potaito) [inline src/gazebo_irlock_plugin.cpp:149]: Is this PR based on an older commit? I added similar changes here, that are not present in this PR: https://github.com/PX4/PX4-SITL_gazebo-classic/pull/919

File: msgs/SITLGps.proto (hunk 3 above)
Comment (JonasPerolini) [reply msgs/SITLGps.proto]: The info is not necessary, I will remove the extra fields

File: src/gazebo_irlock_plugin.cpp (hunk 4 above)
Comment (JonasPerolini) [reply src/gazebo_irlock_plugin.cpp:149]: Yes it was, I've rebased my branch to remove the conflicts. In your PR @potaito you've added the attitude of the target. Here I've added the attitude of the drone when the target is detected.

File: src/gazebo_mavlink_interface.cpp (hunk 2 above)
Comment (JonasPerolini) [reply src/gazebo_mavlink_interface.cpp:967]: The mavlink messages are now merged.

File: src/gazebo_mavlink_interface.cpp [hunk 5]
```diff
@@ -957,6 +962,64 @@ void GazeboMavlinkInterface::IRLockCallback(IRLockPtr& irlock_message) {
   mavlink_interface_->send_mavlink_message(&msg);
 }
 
+void GazeboMavlinkInterface::ArucoMarkerCallback(ArucoMarkerPtr& arucoMarker_message) {
```
Comment (Jaeyoung-Lim) [inline src/gazebo_mavlink_interface.cpp]: Can we keep this a little more generic? e.g. targetReleativeCallback

File: src/gazebo_aruco_plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    // if (sy > 1e-6)
+    // {
+    //     float z = atan2(R.at<double>(1, 0), R.at<double>(0, 0));
+    //     return wrap_2pi(z);
+    // }
+    // else
+    // {
+    //     return 0;
+    // }
+}
```
Comment (Jaeyoung-Lim) [inline src/gazebo_aruco_plugin.cpp]: Missing EOL

File: src/gazebo_irlock_plugin.cpp [hunk 7]
```diff
@@ -80,27 +83,64 @@ void IRLockPlugin::Load(sensors::SensorPtr _sensor, sdf::ElementPtr _sdf)
...
+  ignition::math::Quaterniond& att_W_I = T_W_I.Rot();
+
   gazebo::msgs::LogicalCameraImage img = this->camera->Image();
 
   for (int idx = 0; idx < img.model_size(); idx++) {
 
     gazebo::msgs::LogicalCameraImage_Model model = img.model(idx);
 
-    if (model.has_name() && model.name() == "irlock_beacon") {
+    if (model.has_name() && model.name() == "land_pad") {
```
Comment (Jaeyoung-Lim) [inline src/gazebo_irlock_plugin.cpp]: Maybe better to keep the model name as an sdf parameter? Why do we need to check anyway? We already know which model this plugin is part of. This way we can attach this to any "model" or an "actor"

File: src/gazebo_aruco_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+      arucoMarker_message.set_orientation_q_x(q.X());
+      arucoMarker_message.set_orientation_q_y(q.Y());
+      arucoMarker_message.set_orientation_q_z(q.Z());
+
+      arucoMarker_pub_->Publish(arucoMarker_message);
+    }
+  }
+}
+
+float arucoMarkerPlugin::wrap_2pi(float yaw)
```
Comment (Jaeyoung-Lim) [inline src/gazebo_aruco_plugin.cpp]: Small nitpick
```suggestion
float arucoMarkerPlugin::wrap_2pi(const float &yaw)
```

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]
Sorry for the late review and great work!

PR looks already good as is, but just wasn't sure for a need for a new arucomarker message additional to the IRLock message.

File: src/gazebo_irlock_plugin.cpp (hunk 7 above)
Comment (JonasPerolini) [reply src/gazebo_irlock_plugin.cpp]: From my understanding, the logical camera reports every model in its field of view `gazebo::msgs::LogicalCameraImage_Model model = img.model(idx);`. However, we only want to use the relative pose information if the model is a target.

**Comment by JonasPerolini** [comment]
Thanks for the review @Jaeyoung-Lim I've made the changes that you've requested [here](https://github.com/PX4/PX4-SITL_gazebo-classic/pull/950/commits/21f9a3e8e1e9b8b405da67146f57883d077ff9b0), I'm just not sure how you want to handle [this change](https://github.com/PX4/PX4-SITL_gazebo-classic/pull/950#discussion_r1161091562). Does changing the ArUcoMarker msg to TargetRelative msg works for you?

File: src/gazebo_irlock_plugin.cpp (hunk 7 above)
Comment (Jaeyoung-Lim) [reply src/gazebo_irlock_plugin.cpp]: Okay, could we then at least expose it as a sdf parameter?

**Comment by Jaeyoung-Lim** [comment]
@JonasPerolini That was my thought, unless you see a benefit for having it separate

File: src/gazebo_mavlink_interface.cpp (hunk 5 above)
Comment (JonasPerolini) [reply src/gazebo_mavlink_interface.cpp]: Yes, done in https://github.com/PX4/PX4-SITL_gazebo-classic/pull/950/commits/21f9a3e8e1e9b8b405da67146f57883d077ff9b0

File: src/gazebo_irlock_plugin.cpp (hunk 7 above)
Comment (JonasPerolini) [reply src/gazebo_irlock_plugin.cpp]: Yes, done in https://github.com/PX4/PX4-SITL_gazebo-classic/pull/950/commits/85390e89165d40b3cb1894bee218ec5fc1277fa5

**Review by Jaeyoung-Lim — APPROVED** [review]
LGTM,

Thanks for making the requested changes!

**Comment by Jaeyoung-Lim** [comment]
CI failing due to incompatible mavlink definitions

**Pull request merged**